### ISSUE TITLE
s3: create zero length files to mark directories

### ISSUE THREAD

**Issue by swimfish09** [rclone/rclone#2505]
[store@server2 ~]$ rclone --version
rclone v1.42
- os/arch: linux/amd64
- go version: go1.10.1
command:
rclone copy -vv  /mnt/rbd/ remote:test2 --log-file test.log --stats 3s
when i copy local empty document to ceph rgw, it will not create empty document in ceph rgw

**Comment by ncw** [comment]
I tried copying an empty file to our CEPH luminous cluster both with the swift API and the S3 API and it worked fine for both.

Which version of CEPH are you using?

Can you try

```
touch empty
rclone copy -vv empty remote:test-empty
rclone lsl remote:test-empty
```

What does that report?

**Comment by swimfish09** [comment]
touch is create a file not a document
[store@server0 ~]$ s3cmd la
                       DIR   s3://test/s3backer/
                       DIR   s3://test/empty/
First, I copy s3://test/empty/ to local
And then i will copy local to another bucket
s3://test/empty will not create

**Comment by swimfish09** [comment]
i use CEPH luminous cluster

**Comment by ncw** [comment]
Ah, I see the problem... Either use [the latest beta](https://beta.rclone.org) or remove the `//` in `s3://test/empty` - I think that will work.

**Comment by swimfish09** [comment]
The latest beta may be work, remove the `//` can not work.` s3://test/empty` is just [s3cmd](https://github.com/s3tools/s3cmd) url.

**Comment by ncw** [comment]
> s3://test/empty is just s3cmd url.

Ah yes, I see now!

> The latest beta may be work

I released the 1.43 release over the weekend - does that work?

**Comment by ncw** [comment]
Do you mean creating empty directories not empty files?

Rclone does not support creating empty directories in bucket based remotes such as `s3` or `swift`.

Other tools do this by creating a zero length file object ending with a /, eg `path/to/dir/`.  rclone doesn't do this as, except in the empty directory case, they are a waste of storage and transactions.

Rclone will read those `path/to/dir/` files though.

**Comment by swimfish09** [comment]
Yeah, Other s3 providers do this by creating a zero length file object ending with a /, like Amazon s3.

**Comment by swimfish09** [comment]
So, can rclone give a patch to solve this issue?

**Comment by ncw** [comment]
It could be an option quite easily I suppose.  Do you want to work on it?

I think it would just need the `Mkdir` and `Rmdir` methods to create and delete the objects - everything else should work OK.

**Comment by jacp1** [comment]
There is some discussion in #753 (October 2016) on the details of how other services create empty directories in bucket systems, using the standard would make repos created by rclone interoperable with other tools.

I would like to be able to use rclone to make an offsite mirror of a read-only archive containing projects that have been created by multiple people at the small company where I work. Having to tell people that everything will work except empty directories (which might be important to them), or having to run a script to find all the empty dirs out of the millions of files in the repo and put zero length files in them would make this solution look too hacky to pass scrutiny. As people have said about this issue in other threads, this is currently a show-stopper for us. I appreciate there are commercial tools for doing this, and your priorities may be more around reducing transactions, but all other things being equal I think there is a definite use case for handling these directories!

**Comment by ncw** [comment]
This should be relatively straight forward to implement if someone wanted to have a go.  It should be gated with a flag `--s3-allow-empty-directories`.

**Comment by ncw** [comment]
This is implemented in v1.63 with the flag

      --s3-directory-markers   Upload an empty object with a trailing slash when a new directory is created
